doc/sphinx/source/styleguide.rst

```diff
@@ -1,4 +1,455 @@
+
+    // Washes a load of laundry, changing its state from DIRTY to CLEAN.
+    PetscError WashingMachineWash(WashingMachine* wm, Laundry* load);
+
+A method can perform a task involving the instance and other data provided as
+arguments, as shown above. In this case, it returns a ``PetscErrorCode``
+indicating success or failure. A method can also provide access to data within
+the instance of the class, returning that data:::
+
+    // Returns the cost (in cents) of washing a load.
+    PetscInt WashingMachineCost(WashingMachine* wm);
+
+If you're familiar with contemporary object-oriented programming languages like
+C++ and Java, you can define methods in very similar ways (as long as you don't
+wander too far into inheritance and other "polymorphic" techniques. If it's
+practical, lead the list of parameters with input values, and place output
+parameters at the end.
+
+Polymorphism in C
+^^^^^^^^^^^^^^^^^
+
+**Question: do we use this idea? PETSc does, but it might not be needed in
+TDycore itself.**
+
+Header Files
+============
+
+In general, there should be a header file for each significant type that
+possesses behaviors in TDycore. Header file names are all lowercase.
+In some cases, a single function (unrelated to a type) may occupy a header
+file, and that header file would be named after the function.  In others, a
+header file may contain a set of related functions, and its name should
+concisely reflect the purpose of those functions.
+
+Self-Contained Headers
+----------------------
+
+Header files are self-contained and have a ``.h`` suffix. A "self-contained"
+header file can be included in a translation unit without regard for rules
+relating to the order of its inclusion, or for other headers that are
+"understood" to be included when it is used.
+
+Briefly, a TDycore header file
+* is located in the ``include/tdycore`` subdirectory of the TDycore repo
+* requires header guards
+* should include all the files that it needs
+* should not require any particular symbols to be defined
+
+Header File Location
+--------------------
+
+To make it easier to deploy TDycore as part of a larger application, we place
+most header files in a ``tdycore`` subdirectory within the ``include`` directory
+of the repository. There is a high-level header file called ``tdycore.h`` within
+the ``include`` directory that includes all the basic headers within this
+``tdycore/`` subdirectory.
+
+This means that headers and source files that reference specific TDycore headers
+must include the ``tdycore/`` directory as part of the header file's path. For
+example, if you want to use TDycore's I/O subsystem in a source file, you would
+place the following near the top of the file:::
+
+    #include <tdycore/tdyio.h>
+
+Alternatively, you can rely on the high-level TDycore header to bring in the
+I/O subsystem:::
+
+    #include <tdycore.h>
+
+**Question**: does anyone have an opinion on the use of quotes in headers vs the
```

Comment:
They have different semantics. If you want to find a header that lives in the same directory and has not been added to the header search path (by `-I.` or similar), then use quotes. Otherwise use angle brackets to skip that search of the current directory.